### PR TITLE
[SPARK-23433][SPARK-25250] [CORE] Later created TaskSet should learn about the finished partitions 

### PR DESCRIPTION
**Pull request by Ngone51**
## What changes were proposed in this pull request?

This is an optional solution for #22806 . 

#21131 firstly implement that a previous successful completed task from zombie TaskSetManager could also succeed the active TaskSetManager, which based on an assumption that an active TaskSetManager always exists for that stage when this happen.  But that's not always true as an active TaskSetManager may haven't been created when a previous task succeed, and this is the reason why #22806 hit the issue.

This pr extends #21131 's behavior by adding `stageIdToFinishedPartitions` into TaskSchedulerImpl, which recording the finished partition whenever a task(from zombie or active) succeed. Thus, a later created active TaskSetManager could also learn about the finished partition by looking into `stageIdToFinishedPartitions ` and won't launch any duplicate tasks.

## How was this patch tested?

Add.

### CODE REVIEW

**Comment by Ngone51** [comment]
ping @squito @cloud-fan @pgandhi999 please have a look at this.

**Comment by pgandhi999** [comment]
@Ngone51 I understand that you had a proposal and we were actively discussing on various solutions in the PR #22806 , but however, I have been working on that PR tirelessly for a few months and we still have an ongoing discussion going on there. Any specific reasons as to why did you create your own PR for the same issue? WDYT @squito @cloud-fan ?

**Comment by Ngone51** [comment]
I saw the quite different understanding when I reviewed your updated pr corresponding to my proposal. So, I think it may be better to implement it by myself. And I just want to make the changes clean and easy for people(not only the guys already involved) to review, since it just bases on #21131 and no need to mix in other changes.  I'm sorry if you do care about this and I'm fine for crediting this to you. No offence.

**Comment by pgandhi999** [comment]
ok keeping that aside for now, I was curious to know how did you test the changes? I have tested my changes on a really large dataset(10 TB approx.) where the issue was more easily reproducible and prominent in the event of fetch failures. If you have tested both the issue as well the fix with your change, posting screenshots from the UI might help here.

**Comment by pgandhi999** [comment]
ok to test

**Comment by squito** [comment]
@pgandhi999 I know this can be frustrating, but you shouldn't take offense at this.  you both have been putting a lot of effort in on this and have contributed a ton.  regardless of whose change is actually merged, the discussion and thought from everyone has driven this forward.  sometimes when an idea isn't properly understood in normal english, its just easier to explain an idea by showing it with the actual code.

This is one of the thornier problems I've seen, as you've seen I've made more than one mistake on it myself.

**Comment by squito** [comment]
Jenkins, add to whitelist

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala [hunk 1]
```diff
@@ -208,7 +211,7 @@ private[spark] class TaskSchedulerImpl(
         taskSetsByStageIdAndAttempt.getOrElseUpdate(stage, new HashMap[Int, TaskSetManager])
       stageTaskSets(taskSet.stageAttemptId) = manager
       val conflictingTaskSet = stageTaskSets.exists { case (_, ts) =>
-        ts.taskSet != taskSet && !ts.isZombie
+        ts.taskSet != manager.taskSet && !ts.isZombie
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: why this change?  `taskSet` should be the same as `manager.taskSet`, as `manager` is created from `taskSet` with `val manager = createTaskSetManager(taskSet, maxTaskFailures)`

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala [hunk 2]
```diff
@@ -238,7 +241,13 @@ private[spark] class TaskSchedulerImpl(
   private[scheduler] def createTaskSetManager(
       taskSet: TaskSet,
       maxTaskFailures: Int): TaskSetManager = {
-    new TaskSetManager(this, taskSet, maxTaskFailures, blacklistTrackerOpt)
+    val finishedPartitions =
+      stageIdToFinishedPartitions.getOrElseUpdate(taskSet.stageId, new HashSet[Int])
+    // filter the task which has been finished by previous attempts
+    val tasks = taskSet.tasks.filterNot{ t => finishedPartitions(t.partitionId) }
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: I'd change this comment a bit, to explain why this is necessary, even briefly.  Something like
"The DAGScheduler does not always know all the tasks that have been completed by other tasksets when completing a stage, so we do an extra filter here, while holding the TaskSchedulerImpl lock.  See SPARK-25250 and markPartitionCompletedInAllTaskSets()`

File: core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala [hunk 3]
```diff
@@ -1206,6 +1206,67 @@ class TaskSchedulerImplSuite extends SparkFunSuite with LocalSparkContext with B
     }
   }
 
+  test("successful tasks from previous attempts could be learnt by later active taskset") {
```
Comment (squito) [inline core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala]: this test is good, but could it instead be folded into "Completions in zombie tasksets update status of non-zombie taskset"?  that would make it easier to maintain (if you think there is a reason it needs to be separate, feel free to push back)

File: core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala [hunk 4]
```diff
@@ -1206,6 +1206,67 @@ class TaskSchedulerImplSuite extends SparkFunSuite with LocalSparkContext with B
...
+
+
+    // success task 1.0 , finish partition 1. But now,
+    // no active tsm exists in TaskScheduler for stage0.
+    tsm0.handleSuccessfulTask(tsm0.taskAttempts(1)(0).taskId, result)
+    assert(tsm0.runningTasks === 8)
+    assert(taskScheduler.stageIdToFinishedPartitions(0).contains(1))
+
+    // submit a new taskset with 10 tasks after someone previous task attempt succeed
+    val attempt1 = FakeTask.createTaskSet(10, stageId = 0, stageAttemptId = 1)
```
Comment (squito) [inline core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala]: also be sure to expand the comment here, about why this is an important test case -- it simulates the race between the DAGScheduler event loop and the task-result-getter threads, where a taskset gets submitted with already complete tasks.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala [hunk 5]
```diff
@@ -837,17 +847,31 @@ private[spark] class TaskSchedulerImpl(
...
+   * haven't been created at the time we call this method. And it is possible since the behaviour
+   * of calling on this method and creating active TaskSetManager is from two different threads,
+   * which are "task-result-getter" and "dag-scheduler-event-loop" separately. Consequently, under
+   * this situation, the active TaskSetManager which is created later could not learn about the
+   * finished partitions and keep on launching duplicate tasks, which may lead to job fail for some
+   * severe cases, see SPARK-25250 for details. So, to avoid the problem, we record the finished
+   * partitions for that stage here and exclude the already finished tasks when we creating active
+   * TaskSetManagers later by looking into stageIdToFinishedPartitions. Thus, active TaskSetManager
+   * could be always notified about the finished partitions whether it has been created or not at
+   * the time we call this method.
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: This comment is good, but a little too wordy.  I'd cut it down to something like:

There is also the possibility that the DAGScheduler submits another taskset at the same time as we're marking a task completed here -- that taskset would have a task for a partition that was already completed.  We maintain the set of finished partitions in stageIdToFinishedPartitions, protected by this, so we can filter those tasks when the taskset is submitted.  See SPARK-25250 for more details.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala [hunk 6]
```diff
@@ -96,6 +96,9 @@ private[spark] class TaskSchedulerImpl(
   private[scheduler] val taskIdToTaskSetManager = new ConcurrentHashMap[Long, TaskSetManager]
   val taskIdToExecutorId = new HashMap[Long, String]
 
+  // Protected by `this`
+  private[scheduler] val stageIdToFinishedPartitions = new HashMap[Int, HashSet[Int]]
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: you can use a `HashMap[Int, BitSet]` here to save some memory.

**Review by squito — COMMENTED** [review]
overall I think this approach makes sense, but lets get more input from others.  I did leave some comments on the code, but nothing big.

I will be gone for a week, so I will probably not make any more comments on this for a while.

**Comment by pgandhi999** [comment]
@squito That is indeed a good point and well taken. I also did run @Ngone51 's code for testing and the tests look good. I shall continue to review the PR meanwhile. Thank you.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala (hunk 1 above)
Comment (pgandhi999) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: I have the same question, why use `manager.taskSet` here?

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala [hunk 7]
```diff
@@ -297,6 +306,7 @@ private[spark] class TaskSchedulerImpl(
       taskSetsForStage -= manager.taskSet.stageAttemptId
       if (taskSetsForStage.isEmpty) {
         taskSetsByStageIdAndAttempt -= manager.taskSet.stageId
+        stageIdToFinishedPartitions -= manager.taskSet.stageId
```
Comment (pgandhi999) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: Should we also add this code when calling `killTasks`?

**Comment by srowen** [comment]
I get where you're coming from @pgandhi999 . I'd read this as part of one larger discussion, and indeed it happens sometimes that the easiest way to consider an alternative is to take a look at it concretely. I don't read this as a hijack or anything.

I haven't read the long discussion on the other PR but can see it's a tricky change and you have a couple people with deep experience working on it with you to get it right. It may be that you cross-pollinate ideas between the two PRs. It doesn't so much matter which one gets merged as long as it's a consensus and the best we can come up with.

If the issue is credit then obviously anyone who reads these knows you've driven a lot of the discussion on the important change and this is based on your change.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala (hunk 5 above)
Comment (squito) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: And can you also add a comment that this should only be called with a lock on `this`?

File: core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala (hunk 3 above)
Comment (Ngone51) [reply core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala]: Good idea!

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala (hunk 1 above)
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: Previously, I filter the already completed tasks in `createTaskSetManager()` and creates a new TaskSet by filtered tasks. So, the submitted TaskSet form DAGScheduler is not the same with TaskSet in TaskSetManager. But it failed plenty of unit tests since those tests would keep the original submitted TaskSet for some operations.
And now I have moved the `filter` logic into the TaskSetManager.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala (hunk 5 above)
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: updated.

**Comment by Ngone51** [comment]
@pgandhi999 Thanks for your understanding and helping test this,  appreciate a lot.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala (hunk 7 above)
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: I didn't find method `killTasks` in TaskSchedulerImpl, and for some similar func, e.g. `cancelTasks`, I think it's unnecessary.

**Comment by Ngone51** [comment]
@pgandhi999 @squito Thank you for review, have updated.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala [hunk 8]
```diff
@@ -189,6 +189,18 @@ private[spark] class TaskSetManager(
...
 
+  {
+    // TaskSet got submitted by DAGScheduler may have some already completed
+    // tasks since DAGScheduler does not always know all the tasks that have
+    // been completed by other tasksets when completing a stage, so we mark
+    // those tasks as finished here to avoid launching duplicate tasks, while
+    // holding the TaskSchedulerImpl lock.
+    // See SPARK-25250 and markPartitionCompletedInAllTaskSets()`
+    sched.stageIdToFinishedPartitions
+      .getOrElseUpdate(taskSet.stageId, new BitSet)
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: we can avoid creating empty bit set
```
sched.stageIdToFinishedPartitions.get(taskSet.stageId).foreach {
  finishedPartitions => finishedPartitions.foreach(markPartitionCompleted(_, None))
}
```

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala [hunk 9]
```diff
@@ -837,17 +841,27 @@ private[spark] class TaskSchedulerImpl(
...
+   * stageIdToFinishedPartitions, protected by this, so we can detect those tasks when the taskset
+   * is submitted. See SPARK-25250 for more details.
+   *
+   * note: this method must be called with a lock on this.
    */
   private[scheduler] def markPartitionCompletedInAllTaskSets(
       stageId: Int,
       partitionId: Int,
-      taskInfo: TaskInfo) = {
+      taskInfo: Option[TaskInfo]) = {
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: since no one calls this method with `taskInfo = None`

**Comment by cloud-fan** [comment]
can we address https://github.com/apache/spark/pull/21131#discussion_r203257926 as well?

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala (hunk 8 above)
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: good idea!

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala (hunk 9 above)
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: good catch!

**Comment by Ngone51** [comment]
> can we address [#21131 (comment)](https://github.com/apache/spark/pull/21131#discussion_r203257926) as well?

I revisit the discussion and code, I think maybe we could not remove `maybeFinishTaskSet()` there. Think about the case:

Stage 0 has 5 partitions and there's an active TSM 0.1 and a zombie TSM 0.0 in TaskScheduler currently. TSM 0.1 has finished partition [0, 1, 2, 3] and no running tasks(may be blocked by resource offer or delay scheduling) at the moment a succeed task from TSM 0.0 finished. Thus, we have `successful[4]=true` for TSM 0.1 by calling `markPartitionCompleted`. So, if we do not call `maybeFinishTaskSet` in the `markPartitionCompleted`, TSM 0.1 can not finish at the end because it has no tasks to launch, which results in we could not call `maybeFinishTaskSet` at somewhere else.

WDYT ? @cloud-fan

**Comment by cloud-fan** [comment]
makes sense, since `markPartitionCompleted` is called in more than one place now.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala [hunk 10]
```diff
@@ -189,6 +189,18 @@ private[spark] class TaskSetManager(
...
 
+  {
+    // TaskSet got submitted by DAGScheduler may have some already completed
+    // tasks since DAGScheduler does not always know all the tasks that have
+    // been completed by other tasksets when completing a stage, so we mark
+    // those tasks as finished here to avoid launching duplicate tasks, while
+    // holding the TaskSchedulerImpl lock.
+    // See SPARK-25250 and markPartitionCompletedInAllTaskSets()`
+    sched.stageIdToFinishedPartitions.get(taskSet.stageId).foreach {
+      finishedPartitions => finishedPartitions.foreach(markPartitionCompleted(_, None))
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: How do we guarantee that there is no race condition between the creation of `TaskSetManager` and the updating of `TaskSchedulerImpl. stageIdToFinishedPartitions`?

Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: we always hold lock on `TaskSchedulerImpl` while accessing `stageIdToFinishedPartitions`:

* `TaskSchedulerImpl.submitTasks`

* `TaskSchedulerImpl.handleSuccessfulTask`

* `TaskSchedulerImpl.taskSetFinished`

Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: ah i see, `TaskSetManager` is always created at `* TaskSchedulerImpl.submitTasks`

**Comment by cloud-fan** [comment]
LGTM

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala [hunk 11]
```diff
@@ -292,13 +298,16 @@ private[spark] class TaskSchedulerImpl(
...
-  def taskSetFinished(manager: TaskSetManager): Unit = synchronized {
+  def taskSetFinished(manager: TaskSetManager, success: Boolean): Unit = synchronized {
     taskSetsByStageIdAndAttempt.get(manager.taskSet.stageId).foreach { taskSetsForStage =>
       taskSetsForStage -= manager.taskSet.stageAttemptId
       if (taskSetsForStage.isEmpty) {
         taskSetsByStageIdAndAttempt -= manager.taskSet.stageId
       }
     }
+    if (success) {
+      stageIdToFinishedPartitions -= manager.taskSet.stageId
```
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: I just relaize that we could only remove the finished partitions for the stage when the active TSM succeed.

**Comment by Ngone51** [comment]
Jenkins, retest this please.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala (hunk 11 above)
Comment (pgandhi999) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: Rest all LGTM, had just one question, it it better to remove finished partitions from DAGScheduler Event loop on stage completion to avoid potential race(Suppose one stage attempt is fully complete but the other one is still running till the time it gets killed). Just thinking out loud.

Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: Access to `stageIdToFinishedPartitions` is always under the protection of `TaskSchedulerImpl`, see
https://github.com/apache/spark/pull/23871#discussion_r260593142 . So, I think there won't be any race on it. WDYT ? @pgandhi999

Comment (pgandhi999) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: Ahh I see, makes sense.

**Comment by pgandhi999** [comment]
LGTM

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala [hunk 12]
```diff
@@ -837,19 +846,31 @@ private[spark] class TaskSchedulerImpl(
...
+   * is submitted. See SPARK-25250 for more details.
+   *
+   * note: this method must be called with a lock on this.
    */
   private[scheduler] def markPartitionCompletedInAllTaskSets(
       stageId: Int,
       partitionId: Int,
       taskInfo: TaskInfo) = {
+    // if we do not find a BitSet for this stage, which means an active TaskSetManager
+    // has already succeeded and removed the stage.
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: when can this happen? Or it's just a safe guard?

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala [hunk 13]
```diff
@@ -189,6 +189,18 @@ private[spark] class TaskSetManager(
     addPendingTask(i)
   }
 
+  {
+    // TaskSet got submitted by DAGScheduler may have some already completed
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: shall we do it in `TaskSchedulerImpl.createTaskSetManager`? In general I'd avoid adding too much code to the constructor.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala [hunk 14]
```diff
@@ -530,7 +542,7 @@ private[spark] class TaskSetManager(
 
   private def maybeFinishTaskSet() {
     if (isZombie && runningTasks == 0) {
-      sched.taskSetFinished(this)
+      sched.taskSetFinished(this, tasksSuccessful == numTasks)
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: but you said only active task set manager can clean up the `stageIdToFinishedPartitions`? This branch is under `if (isZombie)`

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala [hunk 15]
```diff
@@ -292,13 +298,16 @@ private[spark] class TaskSchedulerImpl(
    * given TaskSetManager have completed, so state associated with the TaskSetManager should be
    * cleaned up.
    */
-  def taskSetFinished(manager: TaskSetManager): Unit = synchronized {
+  def taskSetFinished(manager: TaskSetManager, success: Boolean): Unit = synchronized {
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: Can't we clean up the `stageIdToFinishedPartitions` when there is no more task set manager for the stage? Then we can avoid adding a new parameter.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala (hunk 13 above)
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: Actually, I have the same thought. Let me try it.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala (hunk 15 above)
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: When a zombie TaskSetManager finished, it could also lead to empty TaskSetManager for the stage(just see `taskSetFinished()`).  So, if we clean up the `stageIdToFinishedPartitions` when there is no more task set manager for the stage, the active TaskSetManager which is created later would lost finished partitions. WDYT ? @cloud-fan

Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: Possibly, we could move the clean up logic into `TaskSetManager.maybeFinishTaskSet()`, then we could avoid adding a new parameter here. But it seems a little weird to clean a TaskSchedulerImpl's stage related variable in TSM rather than TaskSchedulerImpl. WDYT? @cloud-fan

**Comment by Ngone51** [comment]
Jenkins, retest this please.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala [hunk 16]
```diff
@@ -530,7 +530,7 @@ private[spark] class TaskSetManager(
...
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: actually, I don't quite understand why `tasksSuccessful == numTasks` means an active task set manager is finished. How do we know this is an active task set manager?

Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: I think I made a mistake here. Previously I was thinking a zombie TSM must have at least one failed tasks, so tasksSuccessful can't be equal to numTasks. But now I realize that `markPartitionCompleted` could increase `tasksSuccessful` even if it's an already failed task for that zombie TSM. So, it still be possible for a zombie TSM to achieve  `tasksSuccessful == numTasks`. @cloud-fan

Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: Up to my knowledge, we can know a stage is finished in `DAGScheduler.markStageAsFinished`. Maybe we can clean up `stageIdToFinishedPartitions` there?

Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: Clean up `stageIdToFinishedPartitions` in `DAGSchedule.markStageAsFinished` would be definitely work, though it would add an new place to race for `TaskSchedulerImpl` lock. 

Another way is to distinguish TSM active or zombie  in `markPartitionCompleted()` where an active TSM won't be zombie at the time and clean up `stageIdToFinishedPartitions` right there. Which one do you prefer ? @cloud-fan

Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: How do we know a TSM is not zombie in `markPartitionCompleted`?

Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: Oh, sorry, it should be `handleSuccessfulTask()` rather than `markPartitionCompleted()`:

https://github.com/apache/spark/blob/bc7592ba1186001127ecfae327ac22a0727b8bab/core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala#L781-L783

where we firstly mark an active TSM into a zombie one.

**Comment by cloud-fan** [comment]
Let's take a step back and re-understand the initial problem.

My understanding is: a stage can have multiple task set managers with one of them being active. However, the zombie task set managers can still have running tasks. This can lead to two problems:
1. A task from the zombie TSM finishes, later the active TSM launches a new task for the same partition, which keeps failing
2. A task from the zombie TSM finishes. But right before it finishes, the active TSM launches a new task for the same partition. After the task from zombie TSM finishes, the new task keeps failing.

The recent work only focuses on case 1, and doesn't work for case 2 IIUC. Is it the right understanding or I miss something?

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala (hunk 16 above)
Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: `handleSuccessfulTask` SGTM

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala [hunk 17]
```diff
@@ -202,8 +205,11 @@ private[spark] class TaskSchedulerImpl(
     val tasks = taskSet.tasks
     logInfo("Adding task set " + taskSet.id + " with " + tasks.length + " tasks")
     this.synchronized {
-      val manager = createTaskSetManager(taskSet, maxTaskFailures)
       val stage = taskSet.stageId
+      // only create a BitSet once for a certain stage since we only remove
+      // that stage when an active TaskSetManager succeed.
+      stageIdToFinishedPartitions.getOrElseUpdate(stage, new BitSet)
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: nit: can we move it to `createTaskSetManager`, and keep `submitTasks` untouched?

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala [hunk 18]
```diff
@@ -797,11 +801,12 @@ private[spark] class TaskSetManager(
...
   }
 
-  private[scheduler] def markPartitionCompleted(partitionId: Int, taskInfo: TaskInfo): Unit = {
+  private[scheduler] def markPartitionCompleted(partitionId: Int, taskInfo: Option[TaskInfo])
+  : Unit = {
     partitionToIndex.get(partitionId).foreach { index =>
       if (!successful(index)) {
         if (speculationEnabled && !isZombie) {
-          successfulTaskDurations.insert(taskInfo.duration)
+          taskInfo.foreach { info => successfulTaskDurations.insert(info.duration) }
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: this is existing logic, but I have a question here: Do we really need to do it? The task is finished by another TSM, it seems unreasonable to update the statistics for launching speculative tasks in this TSM.

Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: I think you're right. What's your opinion ? @squito

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala (hunk 17 above)
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: updated.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala (hunk 18 above)
Comment (squito) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: Yeah there was discussion about this in the past, there are arguments for doing it multiple ways.  This was kind of a compromise with something that avoided a bug and was a reasonable change to put in, see more here: https://github.com/apache/spark/pull/21656

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala [hunk 19]
```diff
@@ -238,7 +241,20 @@ private[spark] class TaskSchedulerImpl(
...
+    // only create a BitSet once for a certain stage since we only remove
+    // that stage when an active TaskSetManager succeed.
+    stageIdToFinishedPartitions.getOrElseUpdate(taskSet.stageId, new BitSet)
+    val tsm = new TaskSetManager(this, taskSet, maxTaskFailures, blacklistTrackerOpt)
+    // TaskSet got submitted by DAGScheduler may have some already completed
+    // tasks since DAGScheduler does not always know all the tasks that have
+    // been completed by other tasksets when completing a stage, so we mark
+    // those tasks as finished here to avoid launching duplicate tasks, while
+    // holding the TaskSchedulerImpl lock.
+    // See SPARK-25250 and markPartitionCompletedInAllTaskSets()`
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: super nit: run-on sentence and stray '`'.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala [hunk 20]
```diff
@@ -797,16 +801,20 @@ private[spark] class TaskSetManager(
     maybeFinishTaskSet()
   }
 
-  private[scheduler] def markPartitionCompleted(partitionId: Int, taskInfo: TaskInfo): Unit = {
+  private[scheduler] def markPartitionCompleted(partitionId: Int, taskInfo: Option[TaskInfo])
+  : Unit = {
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: super nit: I'm pretty sure that even if just the return type goes over the line, our style is to switch to multi-line definition

File: core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala [hunk 21]
```diff
@@ -1133,30 +1133,51 @@ class TaskSchedulerImplSuite extends SparkFunSuite with LocalSparkContext with B
       assert(tsm.runningTasks === 9)
       tsm
     }
+    // we've now got 2 zombie attempts, each with 9 tasks still active but zero active attempt
+    // in taskScheduler.
```
Comment (squito) [inline core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala]: i don't understand "zero active attempt in taskScheduler", can you explain what you mean?

File: core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala [hunk 22]
```diff
@@ -1133,30 +1133,51 @@ class TaskSchedulerImplSuite extends SparkFunSuite with LocalSparkContext with B
...
+
+    // finish partition 1,2 by completing the tasks before a new attempt for the same stage submit.
+    // And it's possible since the behaviour of submitting new attempt and handling successful task
```
Comment (squito) [inline core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala]: This is possible since ...

File: core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala [hunk 23]
```diff
@@ -1133,30 +1133,51 @@ class TaskSchedulerImplSuite extends SparkFunSuite with LocalSparkContext with B
...
+    // Submit the 3rd attempt still with 10 tasks, this happens due to the race between thread
+    // "task-result-getter" and "dag-scheduler-event-loop", where a TaskSet gets submitted with
+    // already completed tasks. And this time with insufficient resources so not all tasks are
+    // active.
     val finalAttempt = FakeTask.createTaskSet(10, stageAttemptId = 2)
     taskScheduler.submitTasks(finalAttempt)
     val finalTsm = taskScheduler.taskSetManagerForAttempt(0, 2).get
+    // Though, finalTsm gets submitted after some tasks succeeds, but it could also know about the
+    // finished partition by looking into `stageIdToFinishedPartitions` when it is being created,
+    // so that it won't launch any duplicate tasks later.
```
Comment (squito) [inline core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala]: I'd reword this to

Though finalTSM gets submitted with 10 tasks, the call to `taskScheduler.submitTasks` should realize that 2 tasks have already completed, and mark them appropriately, so it won't launch any duplicate tasks later (SPARK-25250).

File: core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala [hunk 24]
```diff
@@ -124,7 +124,8 @@ class FakeTaskScheduler(sc: SparkContext, liveExecutors: (String, String)* /* ex
     }
   }
 
-  override def taskSetFinished(manager: TaskSetManager): Unit = finishedManagers += manager
+  override def taskSetFinished(manager: TaskSetManager)
+  : Unit = finishedManagers += manager
```
Comment (squito) [inline core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala]: if this really needs to be multiline (though I dont' think it does), be sure to use multi-line def and wrap the body in braces

**Review by squito — APPROVED** [review]
Just some very minor style stuff, otherwise lgtm

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala (hunk 18 above)
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: I see.

File: core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala (hunk 21 above)
Comment (Ngone51) [reply core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala]: Hmm, I mean "there's no active attempt in taskScheduler", updated.

**Comment by Ngone51** [comment]
@squito  Thanks for your comments, have updated.

**Review by cloud-fan — APPROVED** [review]

**Review by squito — APPROVED** [review]
lgtm

**Comment by squito** [comment]
merged to master & 2.4.  @Ngone51 there was a merge conflict against branch 2.3, would you like to open another PR for that branch?

Thanks for all the help from everyone on this. @pgandhi999 as a way to split credit, I assigned the jira to you.  definitely appreciate all the work you put in on it

**Comment by pgandhi999** [comment]
Thank you @squito . Once again I appreciate all the work that you guys do and happy to know that we all learned something as a whole by the end of it.

**Comment by Ngone51** [comment]
Thank you @pgandhi999 @squito @cloud-fan @srowen .

I'll submit a pr to help backport to 2.3 later, @squito .
